lib/autoproj/cli/daemon.rb

```diff
@@ -101,67 +101,98 @@ def clear_and_dump_cache
+                    next if buildconf_pull_request?(pull_request)
+
+                    handle_pull_request_modifications(pull_request, events)
+                end
+            end
+
             # @param [Github::PushEvent] push_event
             # @return [void]
-            def handle_mainline_push(push_event)
-                package = watcher.package_affected_by_push_event(push_event)
-                Autoproj.message "Push detected on #{push_event.owner}/"\
-                    "#{push_event.name}, branch: #{push_event.branch} "\
-                    "(remote: #{push_event.head_sha}, local: #{package.head_sha})"
+            def handle_mainline_modifications(packages)
+                packages.each do |pkg, events|
+                    Autoproj.message(
+                        "Push detected on #{pkg.name}, current: #{pkg.head_sha}"
+                    )
+                    events.each do |ev|
+                        Autoproj.message(
+                            "  #{ev.owner}/#{ev.name}, branch: #{ev.branch} "\
+                            "#{ev.head_sha}"
+                        )
+                    end
+                end
 
-                bb.build_mainline_push_event(push_event) unless update_failed?
+                if update_failed?
+                    Autoproj.message "Not triggering build, the last update failed"
+                    Autoproj.message "The daemon will attempt to update the workspace, "\
+                                     "and will trigger a new build if that's successful"
+                else
+                    packages.each do |package, events|
+                        bb.post_mainline_changes(package, events)
+                    end
+                end
 
-                clear_and_dump_cache if buildconf_push?(push_event)
+                buildconf_push = packages.each_value.any? do |events|
+                    events.any? { |ev| buildconf_push?(ev) }
+                end
+                clear_and_dump_cache if buildconf_push
                 restart_and_update
             end
 
             # @param [Github::PullRequest] pull_request
             # @return [void]
-            def handle_pull_request_push(pull_request)
+            def handle_pull_request_modifications(pull_request, events)
+                # Check whether the pull request closed, and if we're aware
+                unless pull_request.open?
+                    if cache.include?(pull_request)
+                        handle_pull_request_closed(pull_request)
+                    end
+                    return
+                end
+
+                unless cache.include?(pull_request)
+                    return handle_pull_request_opened(pull_request)
+                end
+
+                handle_pull_request_changes(pull_request, events)
+            end
+
+            # Handle changes made to an existing, already open, pull request
+            def handle_pull_request_changes(pull_request, _events)
```

Comment:
Why the anused argument?

Reply:
I was keen on trying to get events through to buildbot. The issue I faced here is that the `post` method related to pull requests is also used in the boot code, where there are no events.

Of course, I should just have created one method that processes the events, I see that now :P

Since I don't really have the time to change that now, I just removed the parameter